crates/net/eth-wire/src/types/broadcast.rs

```diff
@@ -529,103 +516,156 @@ impl HandleAnnouncement for NewPooledTransactionHashes66 {
+    pub fn from_raw_data_eth68(data: HashMap<TxHash, V>) -> Self {
+        Self::from_raw_data(data, Some(EthVersion::Eth68))
     }
 
-    /// Returns a new [`ValidAnnouncementData`] wrapper around validated
-    /// [`Eth68`](EthVersion::Eth68) announcement data.
-    pub fn new_eth66(data: HashMap<TxHash, Option<(u8, usize)>>) -> Self {
-        Self::new(data, EthVersion::Eth66)
+    /// Wraps raw data with version [`EthVersion::Eth66`].
+    pub fn from_raw_data_eth66(data: HashMap<TxHash, V>) -> Self {
+        Self::from_raw_data(data, Some(EthVersion::Eth66))
     }
 
-    /// Returns a new [`ValidAnnouncementData`] with empty data from an [`Eth68`](EthVersion::Eth68)
+    /// Returns a new [`PartiallyValidData`] with empty data from an [`Eth68`](EthVersion::Eth68)
     /// announcement.
     pub fn empty_eth68() -> Self {
-        Self::new_eth68(HashMap::new())
+        Self::from_raw_data_eth68(HashMap::new())
     }
 
-    /// Returns a new [`ValidAnnouncementData`] with empty data from an [`Eth66`](EthVersion::Eth66)
+    /// Returns a new [`PartiallyValidData`] with empty data from an [`Eth66`](EthVersion::Eth66)
     /// announcement.
     pub fn empty_eth66() -> Self {
-        Self::new_eth66(HashMap::new())
+        Self::from_raw_data_eth66(HashMap::new())
+    }
+
+    /// Returns the version of the message this data was received in if different versions of the
+    /// message exists, either [`Eth66`](EthVersion::Eth66) or [`Eth68`](EthVersion::Eth68).
+    pub fn msg_version(&self) -> Option<EthVersion> {
+        self.version
     }
 
     /// Destructs returning the validated data.
-    pub fn into_data(self) -> HashMap<TxHash, Option<(u8, usize)>> {
+    pub fn into_data(self) -> HashMap<TxHash, V> {
         self.data
     }
+}
 
+/// Validated data from an announcement that
+#[derive(Debug, Deref, DerefMut, IntoIterator, From)]
+#[from(PartiallyValidData<TxMetadata>)]
+pub struct ValidAnnouncementData {
+    #[deref]
+    #[deref_mut]
+    #[into_iterator]
+    data: HashMap<TxHash, TxMetadata>,
+    version: EthVersion,
+}
+
+hash_map_handle_payload_impl!(ValidAnnouncementData,);
+
+impl ValidAnnouncementData {
     /// Destructs returning only the valid hashes and the announcement message version. Caution! If
     /// this is [`Eth68`](EthVersion::Eth68)announcement data, the metadata must be cached
-    /// before call.
+    /// before call to this method.
     pub fn into_request_hashes(self) -> (RequestTxHashes, EthVersion) {
         let hashes = self.data.into_keys().collect::<Vec<_>>();
 
         (RequestTxHashes::new(hashes), self.version)
     }
-}
 
-impl HandleAnnouncement for ValidAnnouncementData {
-    fn is_empty(&self) -> bool {
-        self.data.is_empty()
-    }
+    /// Conversion from [`PartiallyValidData`] from an announcement. Note! [`PartiallyValidData`]
+    /// from an announcement, should have some [`EthVersion`]. Panics if [`PartiallyValidData`] has
+    /// version set to `None`.
```

Comment:
because the type is used also as output for the first validation pass (check empty + check duplicates) on `PooledTransactions` which don't have a version. this is all in the docs.